core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java

```diff
@@ -760,6 +766,177 @@ protected RexNode removeCorrelationExpr(
+    //              LogicalJoin(condition=[IS NOT DISTINCT FROM($0, $1)], joinType=[left])
+    //                LogicalAggregate(group=[{0}]
+    //                  LogicalProject(DEPTNO=[$0])
+    //                    LogicalTableScan(table=[[scott, DEPT]])
+    //                LogicalAggregate(group=[{0}], EXPR$0=[COUNT()])
+    //                  LogicalProject(DEPTNO=[$7])
+    //                    LogicalFilter(condition=[IS NOT NULL($7)])
+    //                      LogicalTableScan(table=[[scott, EMP]])
+    //
+    // Here we perform an early join, preserving all possible CorVar sets from the outer scope
+    // and their corresponding aggregation results. This ensures that for any row from the left
+    // input of the Correlation, there is always an aggregation result available for join output.
+    //
+    // Implementation based on: Improving Unnesting of Complex Queries
+    // (https://dl.gi.de/server/api/core/bitstreams/c1918e8c-6a87-4da2-930a-bfed289f2388/content)
+    if (rel.getGroupType() == Aggregate.Group.SIMPLE
+        && rel.getGroupSet().isEmpty()
+        && !frame.corDefOutputs.isEmpty()
+        && !parentPropagatesNullValues) {
+      final Pair<CorrelationId, Frame> outerFramePair = requireNonNull(this.frameStack.peek());
+      final Frame outFrame = outerFramePair.right;
+      RexBuilder rexBuilder = relBuilder.getRexBuilder();
+
+      int groupKeySize = (int) corDefOutputs.keySet().stream()
+          .filter(a -> a.corr.equals(outerFramePair.left))
+          .count();
+      List<RelDataTypeField> newRelFields = newRel.getRowType().getFieldList();
+      ImmutableBitSet.Builder corFieldBuilder = ImmutableBitSet.builder();
+
+      // Here we record the mapping between the original index and the new project.
+      // For the count, we map it as `case when x is null then 0 else x`.
+      final Map<Integer, RexNode> newProjectMap = new HashMap<>();
+      final List<RexNode> conditions = new ArrayList<>();
+      for (Map.Entry<CorDef, Integer> corDefOutput : corDefOutputs.entrySet()) {
+        CorDef corDef = corDefOutput.getKey();
+        Integer corIndex = corDefOutput.getValue();
+        if (corDef.corr.equals(outerFramePair.left)) {
+          int newIdx = requireNonNull(outFrame.oldToNewOutputs.get(corDef.field));
+          corFieldBuilder.set(newIdx);
+
+          RelDataType type = outFrame.r.getRowType().getFieldList().get(newIdx).getType();
+          RexNode left = new RexInputRef(corFieldBuilder.cardinality() - 1, type);
+          newProjectMap.put(corIndex + groupKeySize, left);
+          conditions.add(
+              relBuilder.isNotDistinctFrom(left,
+                  new RexInputRef(corIndex + groupKeySize,
+                      newRelFields.get(corIndex).getType())));
+        }
+      }
+
+      ImmutableBitSet groupSet = corFieldBuilder.build();
+      final RelNode join = relBuilder.push(outFrame.r)
```

Comment:
you can annotate the code here with the line from the above plan that is being constructed